src/jit/inline.cpp

```diff
@@ -169,14 +164,135 @@ InlineContext::InlineContext()
     , inlObservation(InlineObservation::CALLEE_UNUSED_INITIAL)
 #ifdef DEBUG
     , inlCallee(nullptr)
+    , inlTreeID(0)
+    , inlSuccess(true)
 #endif
 {
     // Empty
 }
 
+//------------------------------------------------------------------------
+// newRoot: construct an InlineContext for the root method
+//
+// Arguments:
+//   compiler - compiler doing the inlining
+//
+// Return Value:
+//    InlineContext for use as the root context
+//
+// Notes:
+//    We leave inlCode as nullptr here (rather than the IL buffer
+//    address of the root method) to preserve existing behavior, which
+//    is to allow one recursive inline.
```

Comment:
Some background: I believe the intent of the original code was to not allow any recursive inlines at all. However there was no notion of a root context and so the code in `fgIsUnboundedInlineRecursion` would actually permit one recursive inline of the root method. When I first realized this, I added a test case `Jit/opts/Inline/tests/fact.cs' demonstrating this is indeed possible, though it's not all that easy to arrange without resorting to the aggressive inline attribute. When in introduced the root context, I could have fixed this bug by setting inlCode in the root context, but I decided for now to keep the current behavior. Hence the comment.

The recursive inline that is still allowed will indeed get a newSuccess context, and it will set inlCode properly, which will prevent subsequent recursive inlines.